SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,214 @@
+namespace SpatialGDK
+{
+SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const PhysicalWorkerName& InWorkerId, const FSubView& InSubView,
+													   const SpatialVirtualWorkerTranslator* InVirtualWorkerTranslator,
+													   TUniqueFunction<void(EntityComponentUpdate)> InUpdateSender)
 	: WorkerId(InWorkerId)
-	, StaticComponentView(InStaticComponentView)
+	, SubView(&InSubView)
 	, VirtualWorkerTranslator(InVirtualWorkerTranslator)
+	, UpdateSender(MoveTemp(InUpdateSender))
 {
-	check(InStaticComponentView != nullptr);
 	check(InVirtualWorkerTranslator != nullptr);
 }
 
-void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentAdded(const Worker_AddComponentOp& Op)
-{
-	check(HandlesComponent(Op.data.component_id));
-
-	MaybeQueueAclAssignmentRequest(Op.entity_id);
-}
-
-void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentUpdated(const Worker_ComponentUpdateOp& Op)
-{
-	check(HandlesComponent(Op.update.component_id));
-
-	MaybeQueueAclAssignmentRequest(Op.entity_id);
-}
-
-void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentRemoved(const Worker_RemoveComponentOp& Op)
+void SpatialLoadBalanceEnforcer::Advance()
 {
-	check(HandlesComponent(Op.component_id));
-
-	if (AclAssignmentRequestIsQueued(Op.entity_id))
+	const FSubViewDelta& SubViewDelta = SubView->GetViewDelta();
+	for (const EntityDelta& Delta : SubViewDelta.EntityDeltas)
 	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Log,
-			   TEXT("Component %d for entity %lld removed. Can no longer enforce the previous request for this entity."), Op.component_id,
-			   Op.entity_id);
-		AclWriteAuthAssignmentRequests.Remove(Op.entity_id);
-	}
-}
-
-void SpatialLoadBalanceEnforcer::OnEntityRemoved(const Worker_RemoveEntityOp& Op)
-{
-	if (AclAssignmentRequestIsQueued(Op.entity_id))
-	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Entity %lld removed. Can no longer enforce the previous request for this entity."),
-			   Op.entity_id);
-		AclWriteAuthAssignmentRequests.Remove(Op.entity_id);
-	}
-}
-
-void SpatialLoadBalanceEnforcer::OnAclAuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
-{
-	// This class should only be informed of ACL authority changes.
-	check(AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID);
+		bool bRefresh = false;
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				bRefresh |= ApplyComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				bRefresh |= ApplyComponentRefresh(Delta.EntityId, Change.ComponentId, Change.CompleteUpdate.Data);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			bRefresh = true;
+			break;
+		case EntityDelta::REMOVE:
+			DataStore.Remove(Delta.EntityId);
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+			DataStore.Remove(Delta.EntityId);
```

Comment:
Why not just call `ApplyComponentUpdate` here?

Reply:
This means we lost completeness of the entity, but gained it again before we next ticked. In this case, we have to assume any part of the state could have changed, and fully refresh the entity